File: superstitioMain/java/superstitio/customStrings/interFace/WordReplace.java
package superstitio.customStrings.interFace;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WordReplace {
    public String WordOrigin;
    public String WordReplace;

    public WordReplace() {
    }

    public WordReplace(String WordOrigin, String WordReplace) {
        this.WordOrigin = WordOrigin;
        this.WordReplace = WordReplace;
    }

    public static String replaceWord(final String string, List<superstitio.customStrings.interFace.WordReplace> replaceRules) {
        final String[] newString = {string};
        for (superstitio.customStrings.interFace.WordReplace replaceRule : replaceRules) {
            newString[0] = replace(newString[0], replaceRule.WordOrigin, replaceRule.WordReplace);
        }
        return newString[0];
    }

    public static String replace(String StringToReplace, CharSequence target, CharSequence replacement) {
        return Pattern
                .compile(target.toString(), Pattern.LITERAL)
                .matcher(StringToReplace)
                .replaceAll(Matcher.quoteReplacement(replacement.toString()));
    }

    public static String[] replaceWord(String[] strings, List<superstitio.customStrings.interFace.WordReplace> replaceRule) {
        List<String> newStrings = new ArrayList<>();
        if (strings != null)
            for (String string : strings)
                newStrings.add(replaceWord(string, replaceRule));
        return newStrings.toArray(new String[0]);
    }

    public boolean hasNullOrEmpty() {
        return StringSetUtility.isNullOrEmpty(this.WordOrigin) || StringSetUtility.isNullOrEmpty(this.WordReplace);
    }

//    public static WordReplace getMockCardStringWithFlavor() {
//        WordReplace retVal = new WordReplace();
//        retVal.NAME = "[MISSING_TITLE]";
//        retVal.DESCRIPTION = "[MISSING_DESCRIPTION]";
//        retVal.UPGRADE_DESCRIPTION = "[MISSING_DESCRIPTION+]";
//        retVal.EXTENDED_DESCRIPTION = new String[]{"[MISSING_0]", "[MISSING_1]", "[MISSING_2]"};
//        retVal.FLAVOR = "[MISSING_FLAVOR]";
//        return retVal;
//    }
}

File: superstitioMain/java/superstitio/customStrings/interFace/WordReplace.kt
package superstitio.customStrings.interFace

import java.util.regex.Matcher
import java.util.regex.Pattern

class WordReplace
{
    lateinit var WordOrigin: String
    lateinit var WordReplace: String

    constructor()

    constructor(WordOrigin: String?, WordReplace: String?)
    {
        if (WordOrigin == null || WordReplace == null)
            return
        this.WordOrigin = WordOrigin
        this.WordReplace = WordReplace
    }

    fun hasNullOrEmpty(): Boolean
    {
        if (::WordReplace.isInitialized && ::WordOrigin.isInitialized)
            return this.WordOrigin.isEmpty() || this.WordReplace.isEmpty()
        return true
    }

    companion object
    {
        @JvmStatic
        fun replaceWord(string: String?, replaceRules: List<WordReplace>): String
        {
            if (string == null) return ""
            val newString = arrayOf(string)
            for (replaceRule in replaceRules)
            {
                newString[0] = replace(newString[0], replaceRule.WordOrigin, replaceRule.WordReplace)
            }
            return newString[0]
        }

        @JvmStatic
        fun replace(StringToReplace: String, target: CharSequence, replacement: CharSequence): String
        {
            return Pattern
                .compile(target.toString(), Pattern.LITERAL)
                .matcher(StringToReplace)
                .replaceAll(Matcher.quoteReplacement(replacement.toString()))
        }

        @JvmStatic
        fun replaceWord(strings: Array<String>?, replaceRule: List<WordReplace>): Array<String>
        {
            if (strings == null) return emptyArray()
            val newStrings: MutableList<String> = ArrayList()
            for (string in strings)
                newStrings.add(replaceWord(string, replaceRule))
            return newStrings.toTypedArray<String>()
        }
    }
}
